Completely overhaul cross zone and world wide methods in quest API.

KinglyKrab opened the pull request:
- Add quest::crosszonecastspellbycharid(character_id, spell_id) to Perl.
- Add quest::crosszonecastspellbygroupid(group_id, spell_id) to Perl.
- Add quest::crosszonecastspellbyraidid(raid_id, spell_id) to Perl.
- Add quest::crosszonecastspellbyguildid(guild_id, spell_id) to Perl.
- Add quest::crosszonedisabletaskbycharid(character_id, task_id) to Perl.
- Add quest::crosszonedisabletaskbygroupid(group_id, task_id) to Perl.
- Add quest::crosszonedisabletaskbyraidid(raid_id, task_id) to Perl.
- Add quest::crosszonedisabletaskbyguildid(guild_id, task_id) to Perl.
- Add quest::crosszoneenabletaskbycharid(character_id, task_id) to Perl.
- Add quest::crosszoneenabletaskbygroupid(group_id, task_id) to Perl.
- Add quest::crosszoneenabletaskbyraidid(raid_id, task_id) to Perl.
- Add quest::crosszoneenabletaskbyguildid(guild_id, task_id) to Perl.
- Add quest::crosszonefailtaskbycharid(character_id, task_id) to Perl.
- Add quest::crosszonefailtaskbygroupid(group_id, task_id) to Perl.
- Add quest::crosszonefailtaskbyraidid(raid_id, task_id) to Perl.
- Add quest::crosszonefailtaskbyguildid(guild_id, task_id) to Perl.
- Add quest::crosszonemoveinstancebycharid(character_id, instance_id) to Perl.
- Add quest::crosszonemoveinstancebygroupid(group_id, instance_id) to Perl.
- Add quest::crosszonemoveinstancebyraidid(raid_id, instance_id) to Perl.
- Add quest::crosszonemoveinstancebyguildid(guild_id, instance_id) to Perl.
- Add quest::crosszoneremovespellbycharid(character_id, spell_id) to Perl.
- Add quest::crosszoneremovespellbygroupid(group_id, spell_id) to Perl.
- Add quest::crosszoneremovespellbyraidid(raid_id, spell_id) to Perl.
- Add quest::crosszoneremovespellbyguildid(guild_id, spell_id) to Perl.
- Add quest::crosszoneresetactivitybycharid(character_id, task_id, activity_id) to Perl.
- Add quest::crosszoneresetactivitybygroupid(group_id, task_id, activity_id) to Perl.
- Add quest::crosszoneresetactivitybyraidid(raid_id, task_id, activity_id) to Perl.
- Add quest::crosszoneresetactivitybyguildid(guild_id, task_id, activity_id) to Perl.
- Add quest::crosszoneupdateactivitybycharid(character_id, task_id, activity_id, activity_count) to Perl.
- Add quest::crosszoneupdateactivitybygroupid(group_id, task_id, activity_id, activity_count) to Perl.
- Add quest::crosszoneupdateactivitybyraidid(raid_id, task_id, activity_id, activity_count) to Perl.
- Add quest::crosszoneupdateactivitybyguildid(guild_id, task_id, activity_id, activity_count) to Perl.
- Add quest::worldwideassigntask(task_id, enforce_level_requirement) to Perl.
- Add quest::worldwidecastspell(spell_id) to Perl.
- Add quest::worldwidedisabletask(task_id) to Perl.
- Add quest::worldwideenabletask(task_id) to Perl.
- Add quest::worldwidefailtask(task_id) to Perl.
- Add quest::worldwidemessage(type, message) to Perl.
- Add quest::worldwidemove(zone_short_name) to Perl.
- Add quest::worldwidemoveinstance(instance_id) to Perl.
- Add quest::worldwideremovespell(spell_id) to Perl.
- Add quest::worldwideremovetask(task_id) to Perl.
- Add quest::worldwideresetactivity(task_id, activity_id) to Perl.
- Add quest::worldwidesetentityvariableclient(variable_name, variable_value) to Perl.
- Add quest::worldwidesetentityvariablenpc(variable_name, variable_value) to Perl.
- Add quest::worldwidesignalclient(signal) to Perl.
- Add quest::worldwidesignalnpc(signal) to Perl.
- Add quest::worldwideupdateactivity(task_id, activity_id, activity_count) to Perl.
- Add eq.cross_zone_cast_spell_by_char_id(character_id, spell_id) to Lua.
- Add eq.cross_zone_cast_spell_by_group_id(group_id, spell_id) to Lua.
- Add eq.cross_zone_cast_spell_by_raid_id(raid_id, spell_id) to Lua.
- Add eq.cross_zone_cast_spell_by_guild_id(guild_id, spell_id) to Lua.
- Add eq.cross_zone_disable_task_by_char_id(character_id, task_id) to Lua.
- Add eq.cross_zone_disable_task_by_group_id(group_id, task_id) to Lua.
- Add eq.cross_zone_disable_task_by_raid_id(raid_id, task_id) to Lua.
- Add eq.cross_zone_disable_task_by_guild_id(guild_id, task_id) to Lua.
- Add eq.cross_zone_enable_task_by_char_id(character_id, task_id) to Lua.
- Add eq.cross_zone_enable_task_by_group_id(group_id, task_id) to Lua.
- Add eq.cross_zone_enable_task_by_raid_id(raid_id, task_id) to Lua.
- Add eq.cross_zone_enable_task_by_guild_id(guild_id, task_id) to Lua.
- Add eq.cross_zone_fail_task_by_char_id(character_id, task_id) to Lua.
- Add eq.cross_zone_fail_task_by_group_id(group_id, task_id) to Lua.
- Add eq.cross_zone_fail_task_by_raid_id(raid_id, task_id) to Lua.
- Add eq.cross_zone_fail_task_by_guild_id(guild_id, task_id) to Lua.
- Add eq.cross_zone_move_instance_by_char_id(character_id, instance_id) to Lua.
- Add eq.cross_zone_move_instance_by_group_id(group_id, instance_id) to Lua.
- Add eq.cross_zone_move_instance_by_raid_id(raid_id, instance_id) to Lua.
- Add eq.cross_zone_move_instance_by_guild_id(guild_id, instance_id) to Lua.
- Add eq.cross_zone_remove_spell_by_char_id(character_id, spell_id) to Lua.
- Add eq.cross_zone_remove_spell_by_group_id(group_id, spell_id) to Lua.
- Add eq.cross_zone_remove_spell_by_raid_id(raid_id, spell_id) to Lua.
- Add eq.cross_zone_remove_spell_by_guild_id(guild_id, spell_id) to Lua.
- Add eq.cross_zone_reset_activity_by_char_id(character_id, task_id, activity_id) to Lua.
- Add eq.cross_zone_reset_activity_by_group_id(group_id, task_id, activity_id) to Lua.
- Add eq.cross_zone_reset_activity_by_raid_id(raid_id, task_id, activity_id) to Lua.
- Add eq.cross_zone_reset_activity_by_guild_id(guild_id, task_id, activity_id) to Lua.
- Add eq.cross_zone_update_activity_by_char_id(character_id, task_id, activity_id, activity_count) to Lua.
- Add eq.cross_zone_update_activity_by_group_id(group_id, task_id, activity_id, activity_count) to Lua.
- Add eq.cross_zone_update_activity_by_raid_id(raid_id, task_id, activity_id, activity_count) to Lua.
- Add eq.cross_zone_update_activity_by_guild_id(guild_id, task_id, activity_id, activity_count) to Lua.
- Add eq.world_wide_assign_task(task_id, enforce_level_requirement) to Lua.
- Add eq.world_wide_cast_spell(spell_id) to Lua.
- Add eq.world_wide_disable_task(task_id) to Lua.
- Add eq.world_wide_enable_task(task_id) to Lua.
- Add eq.world_wide_fail_task(task_id) to Lua.
- Add eq.world_wide_message(type, message) to Lua.
- Add eq.world_wide_move(zone_short_name) to Lua.
- Add eq.world_wide_move_instance(instance_id) to Lua.
- Add eq.world_wide_remove_spell(spell_id) to Lua.
- Add eq.world_wide_remove_task(task_id) to Lua.
- Add eq.world_wide_reset_activity(task_id, activity_id) to Lua.
- Add eq.world_wide_set_entity_variable_client(variable_name, variable_value) to Lua.
- Add eq.world_wide_set_entity_variable_npc(variable_name, variable_value) to Lua.
- Add eq.world_wide_signal_client(signal) to Lua.
- Add eq.world_wide_signal_npc(signal) to Lua.
- Add eq.world_wide_update_activity(task_id, activity_id, activity_count) to Lua.

Akkadius commented on common/servertalk.h:
```diff
@@ -188,55 +188,114 @@
+#define ServerOP_CZMessageRaid 0x4510
+#define ServerOP_CZMessageGuild 0x4511
+#define ServerOP_CZMovePlayer 0x4512
+#define ServerOP_CZMoveGroup 0x4513
```
This overlaps with `ServerOP_Speech` which is `0x4513`

Akkadius replied:
I'd change the speech opcode for a free range; the rest should be free to use

Akkadius commented on common/servertalk.h:
```diff
@@ -188,55 +188,114 @@
 	QSG_LFGuild_RequestGuildInfo };
 
-#define ServerOP_Speech			0x4513
+#define ServerOP_Speech			0x5500
```
Nice; you grabbed it already, nice work

Akkadius commented on zone/questmgr.cpp:
```diff
@@ -3468,79 +3640,493 @@ void QuestManager::CrossZoneMovePlayerByGuildID(int guild_id, const char *zone_s
+
+void QuestManager::CrossZoneResetActivityByCharID(int character_id, uint32 task_id, int activity_id) {
+	QuestManagerCurrentQuestVars();
+	if (RuleB(TaskSystem, EnableTaskSystem) && initiator && owner) {
```
Remove enable task system checks again

Akkadius commented on zone/worldserver.cpp:
```diff
@@ -2107,66 +2352,447 @@ void WorldServer::HandleMessage(uint16 opcode, const EQ::Net::Packet &p)
+	{
+		CZTaskEnablePlayer_Struct* CZUA = (CZTaskEnablePlayer_Struct*) pack->pBuffer;
+		auto client = entity_list.GetClientByCharID(CZUA->character_id);
+		if (client != 0) {
```
For these checks `if (client)` should be enough

Akkadius commented on zone/worldserver.cpp:
```diff
@@ -2107,66 +2352,447 @@ void WorldServer::HandleMessage(uint16 opcode, const EQ::Net::Packet &p)
+	{
+		CZTaskRemoveGuild_Struct* CZTR = (CZTaskRemoveGuild_Struct*) pack->pBuffer;
+		auto client_list = entity_list.GetClientList();
+		for (auto client : client_list) {
```
Not suggesting a change here; but you can inline `for (auto client : entity_list.GetClientList()) {`

Akkadius commented on zone/worldserver.cpp:
```diff
@@ -1888,189 +1888,434 @@ void WorldServer::HandleMessage(uint16 opcode, const EQ::Net::Packet &p)
+		CZMarqueeGuild_Struct* CZMS = (CZMarqueeGuild_Struct*) pack->pBuffer;
+		auto client_list = entity_list.GetClientList();
+		std::string message = CZMS->message;
+		for (auto client : client_list) {			
```
Everywhere where you are doing this; you need to use `&` as right now you are performing memory copies in these loops

`for (auto &client : client_list) {` will use iterate over the client object in the client list by reference instead of copying the client object for each iteration; this is a big performance implication

Note: search in the source for anywhere you see `for (auto &` and you'll see this referenced almost everywhere and should be done in most cases

KinglyKrab commented:
Added min/max status columns to client methods and changed them to reflect your comments. A minimum status column only wouldn't do anything for stopping it from sending GMs, so we need the max status column as well.

Akkadius commented:
Nice work! Just slinging code!
